jira/client.py

```diff
@@ -2223,59 +2223,23 @@ def _get_user_identifier(self, user: User) -> str:
         """
         return user.accountId if self._is_cloud else user.name
 
-    def _get_user_id(self, user: str | None) -> str | None:
```

Comment:
a benefit of keeping this as a method is that it allows users of the library to override its behaviour should they not like its implementation. I think it would be best to keep it in, but have it also accept `User` objects and additionally also first check if the `JIRA.user()` method returns a valid user before attempting the search with a warning to use the `User` object or a specific string

Reply:
On one hand, I agree with that (what's said both here and https://github.com/pycontribs/jira/pull/1874#pullrequestreview-2237715012), it is a nice QoL feature to have, however on the other, I think it doesn't solve the problem in #1284 and it also is not consistent with the JIRA API, as pointed out in #1855; 

In my eyes, the benefit of removing the internal lookup entirely is that:

- it's consistent with the JIRA API, which returns a 4xx error code if no user is found with the provided data - the library will raise a `JIRAError` for the user to handle;
- it solves the ambiguity of the internal search; I think that the operations (add/remove watchers & assign issue) should be clear cut - either the data provided is either valid and the operation succeeds, or it is not and it will raise an error. Having the library do an internal search takes the decision of how to proceed away from the user.

An elegant solution that offers the best of both worlds would be, IMO, adding an additional boolean flag to each function that allows the users to disable the search if they'd like, ie:
```python
    def assign_issue(self,
                     issue: int | str,
                     assignee: str | None | User,
                     with_lookup: bool = False) -> bool:
        
```

Reply:
I love that idea, will get around to looking into this by the end of the week hopefully!